tests/spec/series_ops/test_containment.py

```diff
@@ -12,6 +12,16 @@
         3 | 301
         4 |
         """,
+    e: """
+          |  i1
+        --+-----
+        1 | 101
+        2 | 201
+        2 | 203
+        3 | 333
+        3 | 334
+        4 |
```

Comment:
I'm afraid I don't see how the approach you've got at the moment is going to work because it's not doing any kind of join or filter on patient_id. I think the only reason it appears to work is because you have distinct values in `e` for each patient. If you to the bottom of `e` a line like:
```
2 | 301
```
Then I think you'll find all the SQL tests will fail (though the in-memory one should be fine).

What makes this problem difficult is that at the point where we're constructing the expression we don't have access to the relevant patient_id column.

It's tempting to try to do something like `lhs.table.c.patient_id` but that won't work in every case because we cant' always guarantee that `lhs` comes from a single table – it might be the result of some case expression over several tables.

What we're really want to say is: use the `patient_id` column of the primary table _whatever that ends up being_ in the final query; but I'm not sure how to implement that in SQLAlchemy.

In very old versions of ehrQL I used this horrible trick of having a "placeholder" ID column which got created here:
https://github.com/opensafely-core/ehrql/blob/58ebdbb97e79c96c33a20b669a1087769b2d88d4/databuilder/query_engines/sqlite.py#L59-L79

And then used the SQLAlchemy `replacement_traverse()` function to replace references to it with a real column later:
https://github.com/opensafely-core/ehrql/blob/58ebdbb97e79c96c33a20b669a1087769b2d88d4/databuilder/query_engines/sqlite.py#L295-L337

That's all pretty nasty though.

Reply:
So in a previous [commit](https://github.com/opensafely-core/ehrql/pull/1762/commits/4f25e0c2f2659aa52101cdcdbdbb975b5e8ab9c8) I used patient_id from `self.population_table`, because the `get_select_query_for_node_domain` [already uses it](https://github.com/opensafely-core/ehrql/blob/main/ehrql/query_engines/base_sql.py#L756) - which also sort-of worked, but returned False for the patient that we expect to get None for.

Reply:
Right, yes, and now I remember why I was trying not to use `exists` - because mssql doesn't like it